### ISSUE TITLE
Proposal: type-safe API for watch and Reflector (or Informer?)

### ISSUE THREAD

**Issue by kosta** [kubernetes-client/javascript#1040]
**tl/dr**: I wrote a [type-safe watch API](https://github.com/understand-ai/js-k8s-iterable-watch/) and would like to merge it in here after review.

I tried to use this client to watch kubernetes events but found it quite cumbersome. E.g. when using

[`listNamespace(..., /*watch:*/ true, ...)`](https://kubernetes-client.github.io/javascript/classes/corev1api.corev1api-1.html#listnamespace)

* Without limit, the request blocks forever, as it tries to read the full response body, but that body is in fact "infinite".
* The types are wrong. This does not return a `V1NamespaceList` but a list of watchevent with a type (e.g. "ADDED") and an object that is the Namespace. But the namespace is not deserialized correctly to a V1Namespace, it's just a the json object.
* To get a full event stream, this needs to be called repeatedly.

A slightly better way is using `k8s.Watch`, [like in this example](https://github.com/kubernetes-client/javascript/blob/master/examples/typescript/watch/watch-example.ts). However, that requires that you know the api path, is callback-based and also does not deserialize correctly, as it cannot know the type to deserialize to. ([Also I cannot find it in the docs?](https://kubernetes-client.github.io/javascript/modules.html))

As I wanted a more typesafe API with proper deserialization, I created [js-k8s-iterable-watch](https://github.com/understand-ai/js-k8s-iterable-watch/). It's kind of like your watch API, but type-safe and deserialized to the correct types using [`ObjectSerializer.deserialize()`](https://github.com/kubernetes-client/javascript/blob/master/src/gen/model/models.ts#L1619). It's being used in production for over a year now and working fine. 

The basic idea is to call e.g. `iterableWatch({ kubeConfig, watch: 'events', log: console.log, timeoutSeconds: 120 })` which returns an `AsyncIterable<WatchEvent<CoreV1Event>>` (correctly deserialized) which can the be iterated over. IterableWatch also does all error-handling and restarts to provide a truly infinite stream.

"Full" code to console.log all kubernetes events:

```javascript
for await (const event of iterableWatch({ kubeConfig, watch: 'events', log: console.log, timeoutSeconds: 120 })) {
  // Note: event is of type @kubernetes/client-node WatchEvent<CoreV1Event>
  console.log(event);
}
```

However, in a way, the current implementation of `iterableWatch` is just a proof of concept. To truly integrate it, it needs to be connected to the code generation to fill out the implementations of [`TypeOf`](https://github.com/understand-ai/js-k8s-iterable-watch/blob/main/src/index.ts#L86), `pathOf` and `typeNameOf`.

Also, some minor issues/improvements need to be taken care of: How do you logging? Just console.log? and implement labelSelector/fieldSelector.

Furthermore, I created a [`Reflector`](https://github.com/understand-ai/js-k8s-iterable-watch/blob/main/src/index.ts#L252) (or is that an Informer?) which is similar as your [`Informer/ListWatch`](https://github.com/kubernetes-client/javascript/blob/5e483cbc6581550f87719f40dd12e2a28e23cbbe/src/cache.ts#L26) (but without the on/off eventing - can be retrofitted though). But I find [my `reflector` code](https://github.com/understand-ai/js-k8s-iterable-watch/blob/main/src/index.ts#L313) much simpler than the [`doneHandler` / `watchHandler`](https://github.com/kubernetes-client/javascript/blob/5e483cbc6581550f87719f40dd12e2a28e23cbbe/src/cache.ts#L144) which can be a little hard to follow (But I'm also missing a few features currently). Maybe, if the iterableWatch api works out, I can fit the missing features and merge this as well.

Soo... what do you think? Are you interested in this API? :)

**Comment by brendandburns** [comment]
Thanks for the interest. I'd be happy to look at a PR which adapts what you have to the existing interfaces. I don't want to have two different ways to do watch/informer stuff, and we need to preserve API compatability for other clients, but if you have ways to expand or improve on things we would very much appreciate the contributions.

**Comment by kosta** [comment]
Hi and sorry for my late reply.

I think adapting to your interface totally makes sense, and I'm all for keeping background compatibility. There should not be two different ways to do watching/informers.

Let's focus on the watch first, once we settled that we can see how that translates to informers.

I fear there is one breaking change that is kind-of required if you want type safety. I think improves the situation a lot, but it's your call, of course.

Current `Watch.watch()` implementation:

```
    public async watch(
        path: string,
        queryParams: any,
        callback: (phase: string, apiObj: any, watchObj?: any) => void,
        done: (err: any) => void,
    ): Promise<any>
```

I see two areas of improvement here:

1. We should return an `AsyncIterable` instead of accepting callback/done. This could be done e.g. using overloading in typescript. I think this is best practice in JS, e.g. all the node read streams are also AsyncIterable etc. Then, "both ways" are possible and it's backwards compatible.

2. Instead of using `any`, we could derive the actual type from the `path`. i.e. if path is `/api/v1/namespaces`, then apiObj is `V1Namespace`.  However there are two complications:
  * Currently, the [apiObj and watchObj are just the result of json.parse](https://github.com/kubernetes-client/javascript/blob/d8442a7d76b1a2c665d83fff96c46517597987ad/src/watch.ts#L128), i.e. they're not the actual "model" types. So e.g. Dates are just string, not Date etc. I think this is the critical point. *If we keep it as-is, there is no type safety. If we fix this, we have a breaking change.* We could of course only fix it for the (new) AsyncIterable case, but that is almost the same as having two APIs just with a weirder interface.
  * To keep api compatibility, we'd need to pass the path which can contain a namespace. [This could get tricky in typescript, but maybe is feasible?](https://www.typescriptlang.org/docs/handbook/2/template-literal-types.html#inference-with-template-literals) If we can get an agreement on the other point, I can try to prototype something.

**Comment by kosta** [comment]
Hi feel free to close if you are not interested in this API. As to make the API type-safe _is_ a breaking change, you would need to decide whether you want that or not.
